contracts/bulk-payments/src/bulk_payments.rs

```diff
@@ -0,0 +1,140 @@
+#![no_std]
+
+use multiversx_sc::derive_imports::*;
+use multiversx_sc::imports::*;
+
+const MAX_USERS_ALLOW: usize = 1000;
+const FOUR_HOURS: u64 = 60 * 60 * 4 * 1_000; // 4 hours
+
+#[type_abi]
+#[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, Clone, PartialEq)]
+pub struct UserAddrTimestamp<M: ManagedTypeApi> {
+    pub addr: ManagedAddress<M>,
+    pub timestamp: u64,
+}
+
+#[multiversx_sc::contract]
+pub trait BulkPayments {
+    #[init]
+    fn init(&self) {}
+
+    #[upgrade]
+    fn upgrade(&self) {}
+
+    #[endpoint(optIn)]
+    fn opt_in(&self) {
+        let caller = self.blockchain().get_caller();
+        require!(
+            self.is_user_whitelisted(caller.clone()),
+            "User not whitelisted"
+        );
+
+        let user_already_opted_in = self.is_user_opted_in(caller.clone());
+
+        // If opted in and still getting rewards
+        if user_already_opted_in && !self.is_timestamp_expired(caller.clone()) {
+            sc_panic!("Already opted in");
+        }
+
+        let number_users_opted_in = self.get_number_users_opted_in();
+        let timestamp = self.blockchain().get_block_timestamp();
+
+        if number_users_opted_in >= MAX_USERS_ALLOW {
+            self.try_clear_first_user_if_timestamp_expired();
+        }
+
+        let deadline_timestamp = timestamp + FOUR_HOURS;
+        self.addr_timestamp(caller.clone()).set(deadline_timestamp);
+        self.opted_in_addrs().insert(caller);
+    }
+
+    fn try_clear_first_user_if_timestamp_expired(&self) {
```

Comment:
Wouldn't it be more efficient to clear multiple entries that expired? Maybe not all of them (to avoid outOfGas errors), but in batches?

Reply:
We want to keep the optIn gas cost at minimum so we'll do only the necessary operations.